client/tests/integration/events/data.rs

```diff
@@ -0,0 +1,130 @@
+use std::{sync::mpsc, thread};
+
+use eyre::Result;
+use iroha_core::{config::Configuration, smartcontracts::wasm};
+use iroha_data_model::{prelude::*, transaction::WasmSmartContract};
+use parity_scale_codec::Encode;
+use test_network::{Peer as TestPeer, *};
+
+use crate::wasm::utils::wasm_template;
+
+fn produce_instructions() -> Vec<Instruction> {
+    let domains: Vec<Domain> = (0..4)
+        .map(|domain_index: usize| Domain::test(&domain_index.to_string()))
+        .collect();
+
+    let registers: [Instruction; 4] = domains
+        .clone()
+        .into_iter()
+        .map(IdentifiableBox::from)
+        .map(RegisterBox::new)
+        .map(Instruction::from)
+        .collect::<Vec<_>>()
+        .try_into()
+        .expect("unreachable");
+
+    vec![
+        // domain "0"
+        // pair
+        //      domain "1"
+        //      if false fail else sequence
+        //          domain "2"
+        //          domain "3"
+        registers[0].clone(),
+        Pair::new::<Instruction, _>(
+            registers[1].clone(),
+            IfInstruction::with_otherwise(
+                false,
+                FailBox::new("unreachable"),
+                SequenceBox::new(vec![registers[2].clone(), registers[3].clone()]),
+            ),
+        )
+        .into(),
+    ]
+}
+
+#[test]
+fn instruction_execution_should_produce_events() -> Result<()> {
+    let instructions = produce_instructions().to_vec().into();
+    transaction_execution_should_produce_events(instructions)
+}
+
+#[test]
+fn wasm_execution_should_produce_events() -> Result<()> {
+    let isi_hex: Vec<String> = produce_instructions()
+        .into_iter()
+        .map(|isi| isi.encode())
+        .map(hex::encode)
+        .collect();
+
+    let mut ptr_offset = 0;
+    let mut isi_calls = String::new();
+    for isi in &isi_hex {
+        let ptr_len = isi.len();
+
+        isi_calls.push_str(&format!(
+            "(call $exec_isi (i32.const {ptr_offset}) (i32.const {ptr_len}))",
+            ptr_offset = ptr_offset / 2,
+            ptr_len = ptr_len / 2,
+        ));
+
+        ptr_offset = ptr_len;
+    }
+
+    let wat = format!(
+        r#"
+        (module
+            {wasm_template}
+
+            ;; Function which starts the smartcontract execution
+            (func (export "{main_fn_name}") (param i32 i32)
+                {isi_calls}
+            )
+        )
+        "#,
```

Comment:
I'm using these small `wat` programs to test wasm execution because it's fast to compile. However, a few integration tests where actual rust smartcontract is compiled to wasm should also be added. I was thinking to spawn an external process, in each test, which would first compile the smartcontract and then the test could execute it. These tests would be very slow I presume so I wouldn't make a lot of them and I don't think it would be good idea to execute them on every pr

Reply:
They are the kinds of tests that we could (and should) run during the release period.

Reply:
that would make most sense